From 094b317da29c27656450496f7ee0c7935f4bb8c8 Mon Sep 17 00:00:00 2001
From: Eoghan Conlon O'Neill <eoghan.conlononeill@chainguard.dev>
Date: Fri, 5 Dec 2025 10:30:48 -0700
Subject: [PATCH] builder: Add Chainguard's distro
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This patch is adapted from Konrad's original.

---

The flux-operator assigns a distro value to an image based on the pull
URL.

Currently, it recognizes a couple of flavours of flux-operator that are
served from either AWS or GitHub. As A consequence it errors for any
unknown registry, including Chaninguard's.

This commit changes that behavior by making the code responsible for
that recognize the "chainguard-private" distro that is related to pull
URLs from within the cgr.dev realm.

Signed-off-by: Konrad Zapałowicz <konrad.zapalowicz@chainguard.dev>
---
 internal/builder/images.go | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/internal/builder/images.go b/internal/builder/images.go
index d82097d..20b2ea8 100644
--- a/internal/builder/images.go
+++ b/internal/builder/images.go
@@ -23,6 +23,7 @@ const (
 	EnterpriseAlpine         string = "enterprise-alpine"
 	EnterpriseDistroless     string = "enterprise-distroless"
 	EnterpriseDistrolessFIPS string = "enterprise-distroless-fips"
+	Chainguard               string = "chainguard"
 )
 
 // ExtractComponentImagesFromObjects extracts the container images from the
@@ -146,8 +147,20 @@ func ExtractComponentImagesWithDigest(srcDir string, opts Options) (images []Com
 			variant = EnterpriseDistroless
 		case "ghcr.io/controlplaneio-fluxcd/distroless-fips":
 			variant = EnterpriseDistrolessFIPS
+		case "cgr.dev/chainguard-private/flux-operator":
+		case "cgr.dev/chainguard-private/flux-operator-fips":
+			variant = Chainguard
 		default:
-			return nil, fmt.Errorf("unsupported registry. consider specifying the distribution variant for registry: %s", registry)
+			// The customers who are entitled to an image have that served from their
+			// org. Similairly, a customized image will have a differnet name. However,
+			// the registry top-level domain remains the same. This final check is
+			// going to catch all images coming from Chainguard realm no matter the
+			// actual org or image name.
+			if strings.Contains(registry, "cgr.dev") {
+				variant = Chainguard
+			} else {
+				return nil, fmt.Errorf("unsupported registry. consider specifying the distribution variant for registry: %s", registry)
+			}
 		}
 	}
 
-- 
2.49.0

